### PR TITLE
[core] Constant can be created as from shared memory

### PR DESCRIPTION
**Pull request by praasz**
### Details:
 - Add Constant constructor to create from shared memory. This constructor optimally can take shared object to take owner ship and release it on Constant end-of-life.
 - Follow up of [this request](https://github.com/openvinotoolkit/openvino.genai/pull/871/files#r1773906684)

### Related PR:
- openvinotoolkit/openvino.genai#871

### Tickets:
 - CVS-153989

### CODE REVIEW

File: src/core/include/openvino/op/constant.hpp [hunk 1]
```diff
@@ -186,6 +186,24 @@ class OPENVINO_API Constant : public Op {
     /// \param data A void* to constant data.
     Constant(const element::Type& type, const Shape& shape, const void* data);
 
+    /// \brief Construct a tensor constant from shared memory.
+    ///
+    /// The constant byte size (defined by type and shape) must be lower than shred memory size, otherwise throw
+    /// exception.
+    /// The Constant can take ownership of shared memory if provided shared object is not null pointer manges memory
+    /// lifetime.
```
Comment (t-jankowski) [inline src/core/include/openvino/op/constant.hpp:194]:
```suggestion
    /// The Constant can take ownership of shared memory if provided shared object is not null and manages memory
    /// lifetime.
```

File: src/core/include/openvino/op/constant.hpp [hunk 2]
```diff
@@ -186,6 +186,23 @@ class OPENVINO_API Constant : public Op {
...
+    ///
+    /// \param type   The element type of the tensor constant.
+    /// \param shape  The shape of the tensor constant.
+    /// \param data   The pointer to shared memory.
+    /// \param size   The byte size of shared memory.
+    /// \param so     The shared object to take it ownership (default: nullptr).
+    Constant(const element::Type& type,
+             const Shape& shape,
+             const void* data,
+             size_t size,
```
Comment (ilya-lavrenov) [inline src/core/include/openvino/op/constant.hpp:203]: maybe `size` can be 0 by default which means it's auto-computed as `ov::shape_size(shape)` ?

BTW, why `size` parameter can be useful at all?

Comment (praasz) [reply src/core/include/openvino/op/constant.hpp:203]: Is for validation purposes. 
If not provided user have to check that provided data are big enough for Constant shape and element type (if not required can be removed OV will be simpler).

Set size to default value or remove it, can make this ctor ambiguous with for copy data: 
```cpp
Constant(const element::Type& type, const Shape& shape, const void* data);
```

Then user will need call Constant like

```
Constant(type, shape, ptr);       -> to copy data
Constant(type, shape, ptr, {});   -> to share data no ownership
Constant(type, shape, ptr, so);   -> to share with ownership
```
Will follow your to keep size as default or remove it.

Comment (ilya-lavrenov) [reply src/core/include/openvino/op/constant.hpp:203]: > If not provided user have to check that provided data are big enough for Constant shape and element type (if not required can be removed OV will be simpler).

I think user need to ensure on its side that `shape` and actual memory size are compatible.

**Review by ilya-lavrenov — APPROVED** [review]
Let's reuse in GenAI once it's merged and available in nightly releases

**Pull request merged**